### ISSUE TITLE
Api versioning errors not thrown when routes tagged with ApiVersions exist but an invalid API version is provided

### ISSUE THREAD

**Issue by fernando-almeida** [Microsoft/aspnet-api-versioning#443]
Hi all,
First of all thanks for the nice work you've done piecing these libraries for abstracting the management of API versioning together.
I've currently hit a snag while tinkering with it and based on some additional debugging effort I've pinpointed the issue to the logic in the file and lines mentioned below.
https://github.com/Microsoft/aspnet-api-versioning/blob/c8d7ff7c4a9012c389a98c36d56e833e240e8346/src/Microsoft.AspNetCore.Mvc.Versioning/Versioning/ApiVersionActionSelector.cs:171-174

I was trying to trigger an InvalidApiVersion errors response by providing a non-existing version, parseable via one of the readers, for an existing action tagged with one or two non-matching ApiVersions attributes.
I was issuing a plain GET request to an API endpoint and the 1st thing I noticed was that the call to
SelectBestCandidate gets 2 candidates one of which does not match the HttpMethodConstraint for being a POST method. I did not try to debug the internal MVC code that I assume that is triggering this call, but I'm assuming that it does not care at that point with filtering actions based on HttpMethodConstraints.
As expected by the logic in SelectBestActions, no match final matches are returned for my request.
The condition finalMatches.Count == 0 is truthful and because of that I'm not getting any error message regarding the invalid API version, but rather an HTTP 404 since no matching route will be found.

This doesn't look like the intended behavior to me, but maybe I'm missing something.
Would appreciate your feedback on this.

**Comment by commonsensesoftware** [comment]
I'll assume that you've read the [error responses](../wiki/Error-Responses) documentation on the wiki. It's a little difficult to imagine the picture with pure text, but from what I am discerning from your description, this _may_ be the expected behavior.

Consider the following controller:

```c#
[ApiVersion("1.0")]
[Route("api/[controller]")]
public class FooController : ControllerBase
{
   [HttpGet]
   public IActionResult Get() => Ok();
}
```

Here's how it is expected to behave with the default configuration:

| Request URL | Status | Description |
| -------------- | ------------- | ------------- |
| `GET api/foo` | 400 | API version unspecified |
| `GET api/foo?api-version=1.0` | 200 | Success |
| `GET api/foo?api-version=2.0` | 400 | API version unsupported |
| `POST api/foo?api-version=1.0` | 405 | Method not supported by API version |
| `GET api/foo?api-version=abc` | 400 | Malformed API version |
| `GET api/foo?api-version=1.0&api-version=2.0` | 400 | Ambiguous API versions |
| `GET api/bar?api-version=1.0` | 404 | No such route for any API version |

In short, any time a route _could_ match, the common error responses will be 400 or 405. 400 indicates the API version is unsupported, while 405 indicates that particular method for that API version is unsupported. This is in lieu of 404 so the client knows the route exists, but not the way it is asking. A 404 is likely to truly mean the route/controller/resource does not exist.

I hope that helps clarify things.

**Comment by fernando-almeida** [comment]
Hi Chris,

Did not had the time to circle back into this again, so I’ll assume it was my misunderstanding on the docs.
Thanks for the detailed clarifications.

**Comment by commonsensesoftware** [comment]
No problem. Not everyone chooses or has time to close out their issues, so I like to give a _ping_ or final opportunity to respond before arbitrarily closing issues.

Again, it's somewhat of a _gray_ area. API Versioning has a concept of _could_ match. This can lead to scenarios where you'll get a 400 instead of 404. If you have more questions or need me to elaborate further I'm happy to.

Thanks
